test/Conversion/CalyxToFSM/lower-while.mlir

```diff
@@ -0,0 +1,130 @@
+// RUN: circt-opt --split-input-file -pass-pipeline='calyx.program(calyx.component(lower-calyx-to-fsm))' %s
+
+// CHECK:  calyx.control {
+// CHECK:   fsm.machine @control() attributes {compiledGroups = [@do_add, @do_add, @do_add, @cond], initialState = "fsm_entry", stateType = i1} {
+// CHECK:     fsm.state "fsm_entry" output {
+// CHECK:     } transitions {
+// CHECK:       fsm.transition @seq_0_cond guard {
+// CHECK:       } action {
+// CHECK:       }
+// CHECK:     }
+// CHECK:     fsm.state "fsm_exit" output {
+// CHECK:     } transitions {
+// CHECK:     }
+// CHECK:     fsm.state "seq_1_while_header" output {
+// CHECK:     } transitions {
+// CHECK:       fsm.transition @fsm_exit guard {
+// CHECK:         %true_0 = hw.constant true
+// CHECK:         %0 = comb.xor %lt_reg.out, %true_0 : i1
+// CHECK:         fsm.return %0
+// CHECK:       } action {
+// CHECK:       }
+// CHECK:       fsm.transition @seq_1_while_if guard {
+// CHECK:         fsm.return %lt_reg.out
+// CHECK:       } action {
+// CHECK:       }
+// CHECK:     }
+// CHECK:     fsm.state "seq_1_while_if" output {
+// CHECK:     } transitions {
+// CHECK:       fsm.transition @seq_1_while_if_else guard {
+// CHECK:         %true_0 = hw.constant true
+// CHECK:         %0 = comb.xor %lt_reg.out, %true_0 : i1
+// CHECK:         fsm.return %0
+// CHECK:       } action {
+// CHECK:       }
+// CHECK:       fsm.transition @seq_1_while_if_then guard {
+// CHECK:         fsm.return %lt_reg.out
+// CHECK:       } action {
+// CHECK:       }
+// CHECK:     }
+// CHECK:     fsm.state "seq_1_while_if_then" output {
+// CHECK:     } transitions {
+// CHECK:       fsm.transition @seq_1_while_if_then_seq_0_do_add guard {
+// CHECK:       } action {
+// CHECK:       }
+// CHECK:     }
+// CHECK:     fsm.state "seq_1_while_if_then_seq_1_do_add" output {
+// CHECK:       fsm.ref @do_add() : () -> ()
+// CHECK:     } transitions {
+// CHECK:       fsm.transition @seq_1_while_header guard {
+// CHECK:       } action {
+// CHECK:       }
+// CHECK:     }
+// CHECK:     fsm.state "seq_1_while_if_then_seq_0_do_add" output {
+// CHECK:       fsm.ref @do_add() : () -> ()
+// CHECK:     } transitions {
+// CHECK:       fsm.transition @seq_1_while_if_then_seq_1_do_add guard {
+// CHECK:       } action {
+// CHECK:       }
+// CHECK:     }
+// CHECK:     fsm.state "seq_1_while_if_else" output {
+// CHECK:     } transitions {
+// CHECK:       fsm.transition @seq_1_while_if_else_seq_0_do_add guard {
+// CHECK:       } action {
+// CHECK:       }
+// CHECK:     }
+// CHECK:     fsm.state "seq_1_while_if_else_seq_0_do_add" output {
+// CHECK:       fsm.ref @do_add() : () -> ()
```

Comment:
In the sketch you showed before, this literally embedded a calyx.enable, but now it is this fsm.ref op. If this approach is going to be scalable, I think you will need to more formally define what this fsm.ref means.

Knowing a bit about what you are going for, it seems like this is a higher-level concept that will later be transformed into a lower-level concept. My understanding is this stands in for a control output port that will be materialized, have a signal driven into it in this state, and hooked up to something in the datapath that is waiting for it (i.e. a calyx.group_go in this case).

If we are going to take the progressive lowering approach here, I think we need to better specify `fsm.ref`. Right now it feels like spooky action at a distance, that is specific to the current implementation.

Reply:
Alternatively, if this is going to be a Calyx specific thing, why not just keep `calyx.enable` embedded within the FSM states, rather than defining this new thing, which feels Calyx specific.

Reply:
I missed this (fairly important) in the PR description: I would much rather prefer to embed `calyx.enable` (or more generally, arbitrary code within the `output` of a state). However, this requires a relaxation of the `calyx.enable` verifier, allowing it to be nested within something other than a Calyx control operation. It's a simple fix, however, it would make the Calyx dialect dependent on the FSM dialect, which might be controversial.

Reply:
addressed in the answer on the main thread.